FileSystemSimulator/FileSystem/FileSystem.cpp

```diff
@@ -134,6 +144,48 @@ bool FileSystem::write(size_t i_index, char* i_mem_area, size_t i_count)
   }
 }
 
+bool FileSystem::read(size_t i_index, char* i_mem_area, size_t i_count)
+{
+	if (i_count == 0)
+		return true;
+
+	OpenFileTable::OFTEntry* entry = oft->getEntry(i_index);
+	if (!entry)
+		return false;
+
+	FileDescriptor fd = getFileDescriptor(entry->fd_index);
+	if (fd.file_size == 0) 
+		return false;
+
+	int local_count;
+	if (i_count > fd.file_size - entry->cur_pos) local_count = fd.file_size - entry->cur_pos;
+	else local_count = i_count;
+
+	while (true) {
+
+		size_t buffer_pos = entry->cur_pos % Sector::BLOCK_SIZE;
+		size_t buffer_space = Sector::BLOCK_SIZE - buffer_pos;
+
+		if (local_count <= buffer_space) {
+			std::memcpy(i_mem_area, entry->buffer + buffer_pos, local_count);
+			entry->cur_pos += local_count;
+			return i_count <= fd.file_size - entry->cur_pos;
+		}
+
+		std::memcpy(i_mem_area, entry->buffer + buffer_pos, buffer_space);
```

Comment:
declare the variable eg `size_t have_written = 0;` and keep it up to date, in order to write data to i_mem_area correctly
`std::memcpy(i_mem_area + have_written, entry->buffer + buffer_pos, buffer_space);`